pkg/pdp/aggregator/jobqueue/queue/queue.go

```diff
@@ -0,0 +1,243 @@
+// Copyright (c) https://github.com/maragudk/goqite
+// https://github.com/maragudk/goqite/blob/6d1bf3c0bcab5a683e0bc7a82a4c76ceac1bbe3f/LICENSE
+//
+// This source code is licensed under the MIT license found in the LICENSE file
+// in the root directory of this source tree, or at:
+// https://opensource.org/licenses/MIT
+
+package queue
+
+import (
+	"context"
+	"database/sql"
+	_ "embed"
+	"errors"
+	"fmt"
+	"time"
+
+	internalsql "github.com/storacha/storage/pkg/pdp/aggregator/jobqueue/internal/sql"
+)
+
+//go:embed schema.sql
+var schema string
+
+// rfc3339Milli is like time.RFC3339Nano, but with millisecond precision, and fractional seconds do not have trailing
+// zeros removed.
+const rfc3339Milli = "2006-01-02T15:04:05.000Z07:00"
+
+type NewOpts struct {
+	DB         *sql.DB
+	MaxReceive int // Max receive count for messages before they cannot be received anymore.
+	Name       string
+	Timeout    time.Duration // Default timeout for messages before they can be re-received.
+}
+
+// New Queue with the given options.
+// Defaults if not given:
+// - Logs are discarded.
+// - Max receive count is 3.
+// - Timeout is five seconds.
+func New(opts NewOpts) (*Queue, error) {
+	if opts.DB == nil {
+		return nil, errors.New("db is required")
+	}
+
+	// TODO(forrest): check if a queue with name already exists and fail if the case.
+	if opts.Name == "" {
+		return nil, errors.New("queue name is required")
+	}
+
+	if opts.MaxReceive < 0 {
+		return nil, errors.New("max receive cannot negative")
+	}
+
+	if opts.MaxReceive == 0 {
+		opts.MaxReceive = 3
+	}
+
+	if opts.Timeout < 0 {
+		return nil, errors.New("timeout cannot be negative")
+	}
+
+	if opts.Timeout == 0 {
+		opts.Timeout = 5 * time.Second
+	}
+
+	return &Queue{
+		db:         opts.DB,
+		name:       opts.Name,
+		maxReceive: opts.MaxReceive,
+		timeout:    opts.Timeout,
+	}, nil
+}
+
+type Queue struct {
+	db         *sql.DB
+	maxReceive int
+	name       string
+	timeout    time.Duration
+}
+
+type ID string
+
+type Message struct {
+	ID       ID
+	Delay    time.Duration
+	Received int
+	Body     []byte
+}
+
+func (q *Queue) MaxReceive() int {
+	return q.maxReceive
+}
+
+func (q *Queue) Timeout() time.Duration {
+	return q.timeout
+}
+
+// Send a Message to the queue with an optional delay.
+func (q *Queue) Send(ctx context.Context, m Message) error {
+	return internalsql.InTx(q.db, func(tx *sql.Tx) error {
+		return q.SendTx(ctx, tx, m)
+	})
+}
+
+// SendTx is like Send, but within an existing transaction.
+func (q *Queue) SendTx(ctx context.Context, tx *sql.Tx, m Message) error {
```

Comment:
do we use all the queue methods? if no should we maintain from where this was copied?

Reply:
There are ~2 methods on `Queue` we don't call here: `Send` and `SendAndGetID`. Also, there are certainly some method that are only called by `Queue` internally that we could unexport. Happy to modify this to make the overall port smaller, and exported methods less. 

> should we maintain from where this was copied

I lean towards no for a couple reasons:
1. I've modified the initial port to include details on how many times a message has been received - allows the worker to provide better viability on task execution.
2. I expect we may move away from sqlite in the future, at which point we'll need to port this anyways. 
3. More generally, I have a hunch further modifications may be needed here (within the jobqueue package, not 'here' specifically) in the future.